components/properties/src/script.rs

```diff
@@ -0,0 +1,207 @@
+    /// and may also indicate Script value, as described for the `trie` field.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub extensions: VarZeroVec<'data, ZeroVecULE<Script>>,
+}
+
+impl<'data> ScriptExtensions<'data> {
+    pub fn try_new(
+        trie: CodePointTrie<'data, ScriptWithExt>,
+        extensions: VarZeroVec<'data, ZeroVecULE<Script>>,
+    ) -> Result<ScriptExtensions<'data>, PropertiesError> {
+        
+        // TODO: do validation here
+
+        Ok(ScriptExtensions { trie, extensions })
+    }
+
+    pub fn get_script_val(&self, code_point: u32) -> Script {
+        let sc_with_ext = self.trie.get(code_point);
+
+        if sc_with_ext.is_other() {
+            let ext_idx = sc_with_ext.0 & SCRIPT_X_SCRIPT_VAL;
+            let scx_val = self.extensions.get(ext_idx as usize);
+            let scx_first_sc = scx_val.and_then(|scx| scx.get(0));
+
+            let default_sc_val = <Script as TrieValue>::DATA_GET_ERROR_VALUE;
+
+            scx_first_sc.unwrap_or(default_sc_val)
+        } else if sc_with_ext.is_common() {
+            Script::Common
+        } else if sc_with_ext.is_inherited() {
+            Script::Inherited
+        } else {
+            let script_val = sc_with_ext.0 & SCRIPT_X_SCRIPT_VAL;
+            Script(script_val)
+        }
+    }
+
+    pub fn get_script_extensions_val(&self, code_point: u32) -> ZeroVec<Script> {
+        let sc_with_ext = self.trie.get(code_point);
+
+        if sc_with_ext.is_other() {
+            let ext_idx = sc_with_ext.0 & SCRIPT_X_SCRIPT_VAL;
+            let ext_subarray = self
+                .extensions
+                .get(ext_idx as usize)
+                .map(|zvule| zvule.as_zerovec())
+                .unwrap_or_default();
+            // In the OTHER case, where the 2 higher-order bits of the
+            // `ScriptWithExt` value in the trie doesn't indicate the Script value,
+            // the Script value is copied/inserted into the first position of the
+            // `extensions` array. So we must remove it to return the actual scx array val.
+            let scx_val_iter = ext_subarray.iter().skip(1);
+            let scx_val: ZeroVec<Script> = ZeroVec::<Script>::from_iter(scx_val_iter);
+            scx_val
+        } else if sc_with_ext.is_common() {
+            let ext_idx = sc_with_ext.0 & SCRIPT_X_SCRIPT_VAL;
+            let scx_val = self
+                .extensions
+                .get(ext_idx as usize)
+                .map(|zvule| zvule.as_zerovec())
+                .unwrap_or_default();
+            scx_val
+        } else if sc_with_ext.is_inherited() {
+            let ext_idx = sc_with_ext.0 & SCRIPT_X_SCRIPT_VAL;
+            let scx_val = self
+                .extensions
+                .get(ext_idx as usize)
+                .map(|zvule| zvule.as_zerovec())
+                .unwrap_or_default();
+            scx_val
+        } else {
+            let scx_vec = [self.get_script_val(code_point)];
```

Comment:
Note: `self.trie.get_ule` actually should return `Option<&T::ULE>`.  You need to convert it into a `&ScriptULE` and then into a `&[ScriptULE]` before you can `unwrap_or_default`.

Reply:
Note: you may be able to convert to `Option<&[ScriptExtULE]>` before `Option<&[ScriptULE]>`.  Do what works.

Reply:
I wrote code to get the types to match up, but it's returning a reference to local data owned by the function <--> I'm still unsure how to go from `&ScriptWithExt::ULE` to `&Script` since `AsULE::from_unaligned` returns an owned value.

Reply:
Sorry.  You can convert from `&ScriptWithExt::ULE` to `&Script::ULE`.  And, in fact, I believe those ULEs are the same.  So you don't actually have to convert anything.

Reply:
Also please remove the unneeded `From` impl.

Reply:
Yes. Changes applied, but I still have that that issue of referencing data that is owned within the body of the function:

```
error[E0515]: cannot return value referencing temporary value
   --> components/properties/src/script.rs:153:13
    |
150 |                 .unwrap_or(&Script::Unknown.as_unaligned());
    |                             ------------------------------ temporary value created here
...
153 |             scx_val
    |             ^^^^^^^ returns a value referencing data owned by the current function
```

I tried more awkward variations but got the same result.

Reply:
The `unwrap_or` can only be on a slice.  Map the reference to a slice first, and then use `unwrap_or`.